### ISSUE TITLE
pre-seed from web browser emulator

### ISSUE THREAD

**Issue by aebrer** [nesbox/TIC-80#2056]
Hello!

I am currently struggling with something. I have a usecase where I produce a generator as a tic-80 cartridge, export it via the web export, and now wish to modify the seed for the prng from the browser. 

I've explored a few different options, but what I've been trying lately is to read in the cart.tic file before it's loaded, search for the `seed=[some number]` line, and replace it, then saving cart.tic

It doesn't work, and I was wondering if anyone could help me figure out why? This is my javascript code "seed_fuckery.js", which I am loading prior to loading the emulator:

`// Read the contents of the cart.tic file using an XHR request
const xhr = new XMLHttpRequest();
xhr.open('GET', 'cart.tic', true);
xhr.responseType = 'arraybuffer';
xhr.onload = function() {
  console.log("Loaded cart.tic");
  // Convert the data to an array of bytes
  const bytes = new Uint8Array(xhr.response);

  // Create a file-like object from the array of bytes
  const file = new Blob([bytes], { type: 'application/wasm' });

  // Read the contents of the file using the FileReader API
  const fileReader = new FileReader();
  fileReader.readAsArrayBuffer(file);

  fileReader.onload = function() {
    // Convert the data to an array of bytes
    const bytes = new Uint8Array(fileReader.result);

    // Iterate over the buffer, starting from the beginning
    let pos = 0;
    while (pos < bytes.length) {
      // Read the 4-byte header of the current chunk
      const header = bytes.slice(pos, pos + 4);
      pos += 4;

      // Extract the bank number and chunk type from the first byte of the header
      const bankNumber = header[0] >> 3;
      const chunkType = header[0] & 0x1F;

      // Extract the size of the chunk from the second and third bytes of the header
      const size = new DataView(header.buffer).getUint16(1, true);

      // Read the chunk data from the buffer
      const chunkData = bytes.slice(pos, pos + size);
      pos += size;

      // If the chunk type is 5 (code), process the chunk data
      if (chunkType === 5) {
        // Convert the chunk data from a buffer to a string
        const code = String.fromCharCode.apply(null, chunkData);

        // Find the line that reads "seed=42069" (or any number)
        const lines = code.split('\n');
        for (let i = 0; i < lines.length; i++) {
          const line = lines[i];
          console.log(line)
          if (line.startsWith('seed=')) {
            // Extract the old seed value
            const oldSeed = parseInt(line.slice(5));
            console.log(`Old seed: ${oldSeed}`);

            // Replace the old seed value with the new seed value
            const newSeed = Math.floor(fxrand()*1000000000);
            lines[i] = `seed=${newSeed}`;
            console.log(`New seed: ${newSeed}`);
          }
        }

        // Convert the modified code back to a buffer
        chunkData = Uint8Array.from(lines.join('\n').split('').map(c => c.charCodeAt(0)));
      }

      // Create a Blob from the modified chunk data and header
      const output = new Blob([header, chunkData], { type: 'application/wasm' });

      // Write the output Blob to the output.tic file using the FileSaver API
      saveAs(output, 'cart.tic');
    }
  }
};`

**Comment by aebrer** [comment]
this is semi-related to https://github.com/nesbox/TIC-80/issues/1698 

Specifically, it would be a different way to solve the seeding issue I am trying to solve. @joshgoebel gave me some advice there to look into hacking the pmem, this was the best I was able to come up with.

**Comment by joshgoebel** [comment]
Use diff file names just to confirm it's using the file you think it is.

**Comment by aebrer** [comment]
my original implementation had two filenames, with the output stream saving as cart.tic; unfortunately the output just doesn't save anything... not to mention I’m not sure this implementation would even work when served over the web?   On Sat, Dec 17, 2022 at 3:35 PM, Josh Goebel ***@***.***> wrote:  
Use diff file names just to confirm it's using the file you think it is.

**Comment by joshgoebel** [comment]
Also, this is a really what I would expect to see. You're not opening a new file for output anywhere and you're not writing all the sections to it unless save as is supposed to be doing that which is very weird naming.  

I also don't see you recalculating the new size of the chunk after you've made changes to it?

**Comment by aebrer** [comment]
Yes, I think it's really fair to say I have no idea what I’m doing, I’m completely out of my element so far.  On Sat, Dec 17, 2022 at 3:40 PM, Josh Goebel ***@***.***> wrote:  
Also, this is a really what I would expect to see. You're not opening a new file for output anywhere and you're not writing all the sections to it unless save as is supposed to be doing that which is very weird naming.
I also don't see you recalculating the new size of the chunk after you've made changes to it?

**Comment by joshgoebel** [comment]
I forgot there isn't a PMEM chunk, but it'd be easier to use any of the other binary chunks and just modify a few bytes than all this code to try and alter strings and put them back together again.  Just give up a sprite, or even add a new chunk if you aren't using all 8 banks of any given chunk.

**Comment by aebrer** [comment]
Yeah since I only need to pass in a single int I can't imagine it taking too much space. I’m not even using sprites so I'd love to try that. Could you possibly give me a little guidance in that direction?   On Sat, Dec 17, 2022 at 3:45 PM, Josh Goebel ***@***.***> wrote:  
I forgot there isn't a PMEM chunk, but it'd be easier to use any of the other binary chunks and just modify a few bytes than all this code to try and alter strings and put them back together again.  Just give up a sprite, or even add a new chunk if you aren't using all 8 banks of any given chunk.

**Comment by joshgoebel** [comment]
You have the right idea, load it, modify it, save it, feed the modified cart to TIC-80, but I don't know the exact code/APIs for all of that off the top of my head... sorry.
